### ISSUE TITLE
Filtering rework

### ISSUE THREAD

**Issue by JosefNemec** [JosefNemec/Playnite#2538]
- add support for exclusion filters
- and/or filter settings per field
- option to not show all fields on filter panel and instead add them dynamically only when needed
- rewrite the entire code to something more maintainable

**Comment by jerrygreen** [comment]
Yeah, filtering out the game that are completed, so I can choose from the rest, - one pretty common filter. I'm surprised there's so much work done, which is wonderful, but yet somehow there's no this feature

**Comment by NovaCuriee** [comment]
I definitely support this. I want to add programs to Playnite, because I have 5 hard drives and it's often hard to locate something when Windows search doesn't help. I _could_ tag every game with "game" and every program with "program" and then use those filters. But either way, I think an exclusion filter would be great, and I'm sure other people have other uses for it aside from what I've said.

As a suggestion, I propose that when right clicking a check box, it will put an X there and exclude whatever you've right clicked. This can solve the problem of needing to cycle through all 3 check box states using only left click, which would be annoying and time consuming. I can't provide any input on what the button would be for fullscreen mode, since I don't use it.

**Comment by jerrygreen** [comment]
History repeats itself, and as in Steam, someone finally thought: «I want to add programs into this thing, too» 😀

Honestly I think this app shouldn’t be used for programs though, by one single reason: you don’t normally need to look at a list of programs and think: «which program I’d like to run to entertain myself today» (unlike with games).

So instead, what you really need for programs is just a search tool most likely (like I do)… Windows Search works bad, indeed. MacOS Spotlight works slightly better in this case though (but still not ideal).

I wait till some new groundbreaking tool comes into the place, a cross-platform one, which will become the new standard. Something that looks as beautiful as Spotlight, or better, but works immediate, instant, and has a bunchful of features, like to execute some function within a program (like create a reminder or something), etc. I knew there’s Raycast, which at least in ads looks similar to what I’d want, but it is in some closed beta, many advertised features unavailable, and it is macOS-only:

https://www.raycast.com/

And it is not open source unfortunately, - which is also important…

Now we need something like VSCode of searching tools. Open source, cool, cross-platform.

**Comment by NovaCuriee** [comment]
Well, I wouldn't exactly be looking at it and wondering which program I should use though. Like I said, it would just be because I can't find stuff. That's what happens with my games as well, which is like 3/4 the reason I use Playnite (the other 1/4 being actually browsing for something to play). It's nice to have everything in one place (instead of having a boring folder with a ton of shortcuts). I was also thinking it would be nice to have filters, so I could easily find audio programs, art programs, etc. 

Anyways, my mention of programs wasn't really the point, I just wanted to give another example of why one may want this feature. I do agree with you though that the usage of Playnite is intended for games, and we may need something else cool for our issue specifically.

**Comment by Prometheus720** [comment]
@jerrygreen 
> And it [Raycast] is not open source unfortunately, - which is also important…

On Windows, try Flow Launcher or Wox. Here are some [other alternatives](https://alternativeto.net/software/flow-launcher/?license=opensource)

**Comment by ThinkSalat** [comment]
I think this is related - but the reason I would want exclusion filters is because I use the grouping feature A LOT. and the obvious and quick fix in terms of that would be to make it so that selected items in the filter list are the only group headers that show up.

For example. I select FPS and Shooter from Genres. If I now go and select to group by Genres, the expected behavior would be that you have two groups: FPS and Shooters. The actual behavior is that it aggregates every genre from every game that includes one of those two genres. It then shows ALL the genres in the list of groupings. THIS I think would be an easy fix that doesn't require any UI changes. That's a quick win and I think would cover a LOT of the reason people want exclusion filters in the first place.

**Comment by xsm2** [comment]
I would love if there could be a way of making a custom filter with a logical expression: an input where you could define your own filter, where you would write something like:  Library:"Steam" Or (Not(Platform:"PC (Windows)") And Status:"Instaled")

That would give users so much control over filters, sure it would not be very intutive nor easy to use if you don't have previous experience with coding, but could be great for power users to use with the filter presets as it would allow for far more complex filters

**Comment by minecrawler** [comment]
Most software with very extensive filtering-capabilities offers simple filters for all users and a query string for power users. JIRA is one example which comes to mind. But afair they all have a rather ugly UI xD

What I'd love to see is a way to create custom filters with query strings, which then show up in the side-bar. This should be the default/only way to create filters. The filters will be pre-populated by a number of useful filters which were created using queries. Normal users won't even see that under the hood all filters use queries and power users will have some examples to start with!

<details>
  <summary>Filters which I use...</summary>

  ![Screenshot of filters](https://user-images.githubusercontent.com/1130708/235758451-2f4fbb7d-e227-45f6-8b09-1702c22aa8a1.png)

</details>

Let's be honest. We usually don't use most of the filters. They are ok-ish for demo purposes, but when we look for a game, the top-left search input is the way to go. The above pic is my every-day view which shows me installed games, and I sometimes switch to the "Plan to Play" completion status to select the next game to install.

Also, the [source code for the filter bar](https://github.com/JosefNemec/Playnite/blob/master/source/Playnite.DesktopApp/Controls/Views/FilterPanel.cs) is in URGENT need of some DRY-ing - so a make-over could remedy multiple things!

Maybe @JosefNemec can give some feedback? Implementing this doesn't seem too hard, but should probably be aligned with the project vision and goals...

**Comment by JosefNemec** [comment]
Filter panel is being completely rewritten in P11. I have nothing more to add for now.

**Comment by ELLIOTTCABLE** [comment]
Just to throw another use-case into the list -

outside of Playnite, I (manually, painfully) maintain two lists on every platform:

1. "Windows-only"
2. "Windows-or-macOS"

My primary workstation is vastly more powerful than I can reasonably justify in a gaming PC; not to mention that it has more screen-space and so on. Where possible, I prefer to game on my work machine, it's simply less work. When I *am* on my gaming PC, and I have a whim to play a game from my owned-games list, it's easy to forget it's available over on my macOS box - so I tend to select games from my 'Windows-only' list when I'm sitting at my Windows machine.

I can't figure out how to access these user-assigned Steam or GOG tags from within Playnite, for one thing; but better yet, I'd love this more-configurable platform to allow me to abandon manually maintaining that entirely - I'd like to be able to simply select "NOT Platform:Macintosh" from the filters.
